packages/jsapi-bootstrap/src/useObjectFetcher.ts

```diff
@@ -0,0 +1,111 @@
+import { VariableDescriptor, VariableDefinition } from '@deephaven/jsapi-types';
+import { useContextOrThrow } from '@deephaven/react-hooks';
+import { createContext } from 'react';
+
+/**
+ * Descriptor for a variable by name. Used when needed to fetch an object.
+ */
+export type NameVariableDescriptor = {
+  /** Type of the variable */
+  type: string;
+  /** Name of the variable */
+  name: string;
+};
+
+/**
+ * Descriptor for a variable by id. Used when needed to fetch an object.
+ */
+export type IdVariableDescriptor = {
+  /** Type of the variable */
+  type: string;
+  /** Id of the variable */
+  id: string;
+};
+
+export function isNameVariableDescriptor(
+  value: unknown
+): value is NameVariableDescriptor {
+  return (
+    typeof value === 'object' &&
+    value != null &&
+    typeof (value as NameVariableDescriptor).type === 'string' &&
+    (value as NameVariableDescriptor).name != null
+  );
+}
+
+export function isIdVariableDescriptor(
+  value: unknown
+): value is IdVariableDescriptor {
+  return (
+    typeof value === 'object' &&
+    value != null &&
+    typeof (value as IdVariableDescriptor).type === 'string' &&
+    (value as IdVariableDescriptor).id != null
+  );
+}
+
+export function isVariableDescriptor(
+  value: unknown
+): value is VariableDescriptor {
+  return isNameVariableDescriptor(value) || isIdVariableDescriptor(value);
+}
+
+/**
+ * Function to fetch an object based on a provided descriptor object
+ */
+export type ObjectFetcher<T = unknown> = (
+  descriptor: VariableDescriptor
+) => Promise<T>;
```

Comment:
TIL! Awesome, thanks!

Reply:
Actually, if I'm doing this, I'm getting an error when providing a value at the context provider:
```
Type '(descriptor: VariableDescriptor) => Promise<unknown>' is not assignable to type 'ObjectFetcher<unknown>'.
  Type 'Promise<unknown>' is not assignable to type 'Promise<TT>'.
    Type 'unknown' is not assignable to type 'TT'.
      'TT' could be instantiated with an arbitrary type which could be unrelated to 'unknown'.ts(2322)
index.d.ts(338, 9): The expected type comes from property 'value' which is declared here on type 'IntrinsicAttributes & ProviderProps<ObjectFetcher<unknown> | null>'
```
Same if I do it without the `TT`. Not sure how to resolve that error...

Reply:
I added a cast in the Context, but that kind of defeats the purpose of doing this. Any idea how to get rid of that cast?

Reply:
What type did you try? Here's a [TS playground](https://www.typescriptlang.org/play?ts=4.9.5#code/C4TwDgpgBA8gRgKwgY2AMQsZALCAnAHgBUoBeKAVwDsBrKgewHcqA+MqYkiAD2AioAmAZyglyRFgAoAlGTYAFPPQC2ASyEROLANwAoXQDNqqVfSqUN8JKgxZchMZVoNmU6QC5YiFOkw58xGwA3rpQYVB4mBR45pIyclAGAIYANhqySSLUdEzmmV7WvnYBEnoAvvq6yGZCwFBJnlY+tv547HGypGzJaRAZWc659SJNNn72eroA9FNQiirqmrV4qlQA5ixVNXVw7EkEy6sbMpPVVLVQyOwUlt5jxYRUFMpw+FAAPlCH626TM3NKNQaAhPF5vT7fDZbc51ATsZAnfT-eZApbAFY-aEXaDkZAHdFHX5I2YAUTwSjacCmVzg9HoKQgSXM6igDDqSS+FDgoEgUHoBlZz1ebQhBPWWLqAtxBFp9MZrERQA) that seems to be working on our version of TS

I added an `extends T` to the function generic in the example. This makes the last line an error. If we want that to not be an error, then remove the `extends T`

Reply:
Oh you actually need to cast/mark the return type a few lines above the provider where you create the actual fetcher. Or when we switch to the generated jsapi types it will work b/c `getObject` returns `Promise<any>` in those and that will satisfy it w/ the inferred typings

So cast where you create it or change the jsapi type to return `Promise<any>` instead of `Promise<unknown>`

Reply:
I changed in dh.types.ts